File: solutions/java/0106.java
class Solution {
  public TreeNode buildTree(int[] inorder, int[] postorder) {
    Map<Integer, Integer> map = new HashMap<>();
    for (int i = 0; i < inorder.length; ++i)
      map.put(inorder[i], i);

    return helper(inorder, 0, inorder.length - 1, postorder, 0, postorder.length - 1, map);
  }

  private TreeNode helper(int[] inorder, int iLeft, int iRight, int[] postorder, int pLeft, int pRight,
      Map<Integer, Integer> map) {
    if (iLeft > iRight || pLeft > pRight)
      return null;

    int i = map.get(postorder[pRight]);
    TreeNode curr = new TreeNode(postorder[pRight]);
    curr.left = helper(inorder, iLeft, i - 1, postorder, pLeft, pLeft + i - iLeft - 1, map);
    curr.right = helper(inorder, i + 1, iRight, postorder, pLeft + i - iLeft, pRight - 1, map);

    return curr;
  }
}

File: solutions/python3/0106.py
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        def helper(iLeft: int, iRight: int, pLeft: int, pRight: int) -> TreeNode:
            if iLeft > iRight or pLeft > pRight:
                return None

            i = dict[postorder[pRight]]
            curr = TreeNode(postorder[pRight])
            curr.left = helper(iLeft, i - 1, pLeft, pLeft + i - iLeft - 1)
            curr.right = helper(i + 1, iRight, pLeft + i - iLeft, pRight - 1)

            return curr

        dict = {val: i for i, val in enumerate(inorder)}

        return helper(0, len(inorder) - 1, 0, len(postorder) - 1)
